File: scenarios/basic/cross-platform/qt/KeyboardTrapInWidget.cpp
#include <QApplication>
#include <QMainWindow>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QWidget>
#include <QLineEdit>
#include <QComboBox>
#include <QCalendarWidget>
#include <QDateEdit>
#include <QMessageBox>

class DatePickerWidget : public QMainWindow
{
    Q_OBJECT

public:
    DatePickerWidget(QWidget *parent = nullptr) : QMainWindow(parent)
    {
        setupUI();
        setupConnections();
    }

private slots:
    void showDatePicker()
    {
        datePickerDialog->show();
        datePickerDialog->raise();
        datePickerDialog->activateWindow();
    }
    
    void hideDatePicker()
    {
        datePickerDialog->hide();
    }
    
    void confirmDate()
    {
        QDate selectedDate = datePickerDialog->selectedDate();
        if (selectedDate.isValid()) {
            dateInput->setText(selectedDate.toString("yyyy-MM-dd"));
            datePickerDialog->hide();
        }
    }
    
    void cancelDateSelection()
    {
        datePickerDialog->hide();
    }

private:
    void setupUI()
    {
        QWidget *centralWidget = new QWidget(this);
        setCentralWidget(centralWidget);
        
        QVBoxLayout *mainLayout = new QVBoxLayout(centralWidget);
        
        // Header
        QHBoxLayout *headerLayout = new QHBoxLayout();
        
        QLabel *titleLabel = new QLabel("Event Scheduler", this);
        titleLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 28px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    padding: 20px 0;"
            "}"
        );
        headerLayout->addWidget(titleLabel);
        
        headerLayout->addStretch();
        
        QPushButton *helpBtn = new QPushButton("❓ Help", this);
        helpBtn->setStyleSheet(
            "QPushButton {"
            "    background: #6c757d;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 8px 16px;"
            "    font-size: 14px;"
            "    font-weight: bold;"
            "    margin-right: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #5a6268;"
            "}"
        );
        headerLayout->addWidget(helpBtn);
        
        QPushButton *settingsBtn = new QPushButton("⚙️ Settings", this);
        settingsBtn->setStyleSheet(
            "QPushButton {"
            "    background: #007bff;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 8px 16px;"
            "    font-size: 14px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #0056b3;"
            "}"
        );
        headerLayout->addWidget(settingsBtn);
        
        mainLayout->addLayout(headerLayout);
        
        // Form content
        QHBoxLayout *formLayout = new QHBoxLayout();
        
        // Left column
        QVBoxLayout *leftColumn = new QVBoxLayout();
        
        QLabel *formTitle = new QLabel("Create New Event", this);
        formTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 24px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 20px;"
            "}"
        );
        leftColumn->addWidget(formTitle);
        
        // Event title field
        QHBoxLayout *titleLayout = new QHBoxLayout();
        
        QLabel *titleLabel2 = new QLabel("Event Title *", this);
        titleLabel2->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-right: 15px;"
            "    min-width: 120px;"
            "}"
        );
        titleLayout->addWidget(titleLabel2);
        
        titleInput = new QLineEdit(this);
        titleInput->setPlaceholderText("Enter event title");
        titleInput->setStyleSheet(
            "QLineEdit {"
            "    padding: 12px;"
            "    border: 2px solid #ddd;"
            "    border-radius: 6px;"
            "    font-size: 16px;"
            "    min-width: 300px;"
            "}"
            "QLineEdit:focus {"
            "    border-color: #007bff;"
            "}"
        );
        titleLayout->addWidget(titleInput);
        
        leftColumn->addLayout(titleLayout);
        
        // Date field
        QHBoxLayout *dateLayout = new QHBoxLayout();
        
        QLabel *dateLabel = new QLabel("Event Date *", this);
        dateLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-right: 15px;"
            "    min-width: 120px;"
            "}"
        );
        dateLayout->addWidget(dateLabel);
        
        dateInput = new QLineEdit(this);
        dateInput->setPlaceholderText("Click to select date");
        dateInput->setReadOnly(true);
        dateInput->setStyleSheet(
            "QLineEdit {"
            "    padding: 12px;"
            "    border: 2px solid #ddd;"
            "    border-radius: 6px;"
            "    font-size: 16px;"
            "    min-width: 300px;"
            "    background: #f8f9fa;"
            "}"
            "QLineEdit:focus {"
            "    border-color: #007bff;"
            "}"
        );
        dateLayout->addWidget(dateInput);
        
        // KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        QPushButton *datePickerBtn = new QPushButton("📅", this);
        datePickerBtn->setStyleSheet(
            "QPushButton {"
            "    background: #007bff;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 12px 16px;"
            "    font-size: 16px;"
            "    margin-left: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #0056b3;"
            "}"
        );
        connect(datePickerBtn, &QPushButton::clicked, this, &DatePickerWidget::showDatePicker);
        dateLayout->addWidget(datePickerBtn);
        
        leftColumn->addLayout(dateLayout);
        
        // Time field
        QHBoxLayout *timeLayout = new QHBoxLayout();
        
        QLabel *timeLabel = new QLabel("Event Time", this);
        timeLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-right: 15px;"
            "    min-width: 120px;"
            "}"
        );
        timeLayout->addWidget(timeLabel);
        
        timeInput = new QLineEdit(this);
        timeInput->setPlaceholderText("Enter time (e.g., 14:30)");
        timeInput->setStyleSheet(
            "QLineEdit {"
            "    padding: 12px;"
            "    border: 2px solid #ddd;"
            "    border-radius: 6px;"
            "    font-size: 16px;"
            "    min-width: 300px;"
            "}"
            "QLineEdit:focus {"
            "    border-color: #007bff;"
            "}"
        );
        timeLayout->addWidget(timeInput);
        
        leftColumn->addLayout(timeLayout);
        
        // Location field
        QHBoxLayout *locationLayout = new QHBoxLayout();
        
        QLabel *locationLabel = new QLabel("Location", this);
        locationLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-right: 15px;"
            "    min-width: 120px;"
            "}"
        );
        locationLayout->addWidget(locationLabel);
        
        locationInput = new QLineEdit(this);
        locationInput->setPlaceholderText("Enter event location");
        locationInput->setStyleSheet(
            "QLineEdit {"
            "    padding: 12px;"
            "    border: 2px solid #ddd;"
            "    border-radius: 6px;"
            "    font-size: 16px;"
            "    min-width: 300px;"
            "}"
            "QLineEdit:focus {"
            "    border-color: #007bff;"
            "}"
        );
        locationLayout->addWidget(locationInput);
        
        leftColumn->addLayout(locationLayout);
        
        // Description field
        QHBoxLayout *descriptionLayout = new QHBoxLayout();
        
        QLabel *descriptionLabel = new QLabel("Description", this);
        descriptionLabel->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-right: 15px;"
            "    min-width: 120px;"
            "}"
        );
        descriptionLayout->addWidget(descriptionLabel);
        
        descriptionInput = new QLineEdit(this);
        descriptionInput->setPlaceholderText("Enter event description");
        descriptionInput->setStyleSheet(
            "QLineEdit {"
            "    padding: 12px;"
            "    border: 2px solid #ddd;"
            "    border-radius: 6px;"
            "    font-size: 16px;"
            "    min-width: 300px;"
            "}"
            "QLineEdit:focus {"
            "    border-color: #007bff;"
            "}"
        );
        descriptionLayout->addWidget(descriptionInput);
        
        leftColumn->addLayout(descriptionLayout);
        
        formLayout->addLayout(leftColumn);
        
        // Right column
        QVBoxLayout *rightColumn = new QVBoxLayout();
        
        QLabel *infoTitle = new QLabel("Event Information", this);
        infoTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 20px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 20px;"
            "}"
        );
        rightColumn->addWidget(infoTitle);
        
        QLabel *infoText = new QLabel(
            "Create and manage your events with our easy-to-use scheduler. "
            "Set dates, times, locations, and descriptions for all your important events.",
            this
        );
        infoText->setStyleSheet(
            "QLabel {"
            "    font-size: 14px;"
            "    color: #666;"
            "    line-height: 1.6;"
            "    margin-bottom: 30px;"
            "}"
        );
        infoText->setWordWrap(true);
        rightColumn->addWidget(infoText);
        
        QLabel *featuresTitle = new QLabel("Features", this);
        featuresTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 18px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 15px;"
            "}"
        );
        rightColumn->addWidget(featuresTitle);
        
        QLabel *featuresList = new QLabel(
            "• Easy date selection\n"
            "• Time management\n"
            "• Location tracking\n"
            "• Event descriptions\n"
            "• Calendar integration",
            this
        );
        featuresList->setStyleSheet(
            "QLabel {"
            "    font-size: 16px;"
            "    color: #333;"
            "    line-height: 1.6;"
            "    margin-bottom: 30px;"
            "}"
        );
        rightColumn->addWidget(featuresList);
        
        rightColumn->addStretch();
        formLayout->addLayout(rightColumn);
        
        mainLayout->addLayout(formLayout);
        
        // Action buttons
        QHBoxLayout *buttonLayout = new QHBoxLayout();
        
        QPushButton *clearBtn = new QPushButton("Clear Form", this);
        clearBtn->setStyleSheet(
            "QPushButton {"
            "    background: #6c757d;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 12px 24px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    margin-right: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #5a6268;"
            "}"
        );
        connect(clearBtn, &QPushButton::clicked, [this]() {
            titleInput->clear();
            dateInput->clear();
            timeInput->clear();
            locationInput->clear();
            descriptionInput->clear();
        });
        buttonLayout->addWidget(clearBtn);
        
        buttonLayout->addStretch();
        
        QPushButton *saveBtn = new QPushButton("Save Event", this);
        saveBtn->setStyleSheet(
            "QPushButton {"
            "    background: #007bff;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 12px 24px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #0056b3;"
            "}"
        );
        connect(saveBtn, &QPushButton::clicked, [this]() {
            QMessageBox::information(this, "Success", "Event saved successfully!");
        });
        buttonLayout->addWidget(saveBtn);
        
        mainLayout->addLayout(buttonLayout);
        
        // KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        setupDatePickerDialog();
    }
    
    void setupDatePickerDialog()
    {
        // KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        datePickerDialog = new QWidget(this, Qt::Dialog | Qt::FramelessWindowHint);
        datePickerDialog->setWindowModality(Qt::ApplicationModal);
        datePickerDialog->setStyleSheet(
            "QWidget {"
            "    background: white;"
            "    border: 2px solid #007bff;"
            "    border-radius: 8px;"
            "    padding: 20px;"
            "}"
        );
        
        QVBoxLayout *dialogLayout = new QVBoxLayout(datePickerDialog);
        
        QHBoxLayout *dialogHeaderLayout = new QHBoxLayout();
        
        QLabel *dialogTitle = new QLabel("Select Date", this);
        dialogTitle->setStyleSheet(
            "QLabel {"
            "    font-size: 18px;"
            "    font-weight: bold;"
            "    color: #333;"
            "    margin-bottom: 15px;"
            "}"
        );
        dialogHeaderLayout->addWidget(dialogTitle);
        
        dialogHeaderLayout->addStretch();
        
        QPushButton *closeBtn = new QPushButton("✕", this);
        closeBtn->setStyleSheet(
            "QPushButton {"
            "    background: #dc3545;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 4px;"
            "    padding: 8px 12px;"
            "    font-size: 14px;"
            "    font-weight: bold;"
            "    width: 30px;"
            "    height: 30px;"
            "}"
            "QPushButton:hover {"
            "    background: #c82333;"
            "}"
        );
        connect(closeBtn, &QPushButton::clicked, this, &DatePickerWidget::cancelDateSelection);
        dialogHeaderLayout->addWidget(closeBtn);
        
        dialogLayout->addLayout(dialogHeaderLayout);
        
        // KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        datePicker = new QCalendarWidget(this);
        datePicker->setStyleSheet(
            "QCalendarWidget {"
            "    background: white;"
            "    border: 1px solid #ddd;"
            "    border-radius: 6px;"
            "    font-size: 14px;"
            "}"
            "QCalendarWidget QWidget {"
            "    alternate-background-color: #f8f9fa;"
            "}"
            "QCalendarWidget QAbstractItemView:enabled {"
            "    selection-background-color: #007bff;"
            "    selection-color: white;"
            "}"
        );
        dialogLayout->addWidget(datePicker);
        
        QHBoxLayout *dialogButtonLayout = new QHBoxLayout();
        
        QPushButton *cancelBtn = new QPushButton("Cancel", this);
        cancelBtn->setStyleSheet(
            "QPushButton {"
            "    background: #6c757d;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 10px 20px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "    margin-right: 10px;"
            "}"
            "QPushButton:hover {"
            "    background: #5a6268;"
            "}"
        );
        connect(cancelBtn, &QPushButton::clicked, this, &DatePickerWidget::cancelDateSelection);
        dialogButtonLayout->addWidget(cancelBtn);
        
        dialogButtonLayout->addStretch();
        
        QPushButton *confirmBtn = new QPushButton("Confirm", this);
        confirmBtn->setStyleSheet(
            "QPushButton {"
            "    background: #007bff;"
            "    color: white;"
            "    border: none;"
            "    border-radius: 6px;"
            "    padding: 10px 20px;"
            "    font-size: 16px;"
            "    font-weight: bold;"
            "}"
            "QPushButton:hover {"
            "    background: #0056b3;"
            "}"
        );
        connect(confirmBtn, &QPushButton::clicked, this, &DatePickerWidget::confirmDate);
        dialogButtonLayout->addWidget(confirmBtn);
        
        dialogLayout->addLayout(dialogButtonLayout);
        
        datePickerDialog->hide();
    }
    
    void setupConnections()
    {
        // Additional connections can be added here
    }

private:
    QLineEdit *titleInput;
    QLineEdit *dateInput;
    QLineEdit *timeInput;
    QLineEdit *locationInput;
    QLineEdit *descriptionInput;
    QWidget *datePickerDialog;
    QCalendarWidget *datePicker;
};

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
    
    DatePickerWidget app;
    app.setWindowTitle("Event Scheduler - Keyboard Trap");
    app.resize(800, 600);
    app.show();
    
    return app.exec();
}

#include "KeyboardTrapInWidget.moc"

File: scenarios/basic/desktop/linux-qt/KeyboardTrapInWidget.py
import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QVBoxLayout, QHBoxLayout, 
                             QWidget, QLabel, QPushButton, QFrame, QLineEdit, QTextEdit, 
                             QCalendarWidget, QSpinBox, QMessageBox)
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QFont

class KeyboardTrapInWidget(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Event Scheduler")
        self.setGeometry(100, 100, 800, 600)
        
        # Central widget
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        
        # Main layout
        main_layout = QVBoxLayout(central_widget)
        main_layout.setSpacing(20)
        main_layout.setContentsMargins(20, 20, 20, 20)
        
        # Header
        header_layout = QVBoxLayout()
        header_layout.setAlignment(Qt.AlignCenter)
        
        title_label = QLabel("Event Scheduler")
        title_font = QFont()
        title_font.setPointSize(24)
        title_font.setBold(True)
        title_label.setFont(title_font)
        title_label.setAlignment(Qt.AlignCenter)
        header_layout.addWidget(title_label)
        
        subtitle_label = QLabel("Create and manage your events")
        subtitle_font = QFont()
        subtitle_font.setPointSize(16)
        subtitle_label.setFont(subtitle_font)
        subtitle_label.setAlignment(Qt.AlignCenter)
        header_layout.addWidget(subtitle_label)
        
        main_layout.addLayout(header_layout)
        
        # Event form
        form_frame = QFrame()
        form_frame.setFrameStyle(QFrame.Box)
        form_frame.setStyleSheet("QFrame { background-color: #f8f9fa; }")
        
        form_layout = QVBoxLayout(form_frame)
        form_layout.setSpacing(20)
        form_layout.setContentsMargins(20, 20, 20, 20)
        
        # Event title
        title_layout = QVBoxLayout()
        
        title_label = QLabel("Event Title")
        title_font = QFont()
        title_font.setPointSize(16)
        title_font.setBold(True)
        title_label.setFont(title_font)
        title_layout.addWidget(title_label)
        
        self.title_entry = QLineEdit()
        self.title_entry.setPlaceholderText("Enter event title")
        self.title_entry.setFixedSize(300, 35)
        title_layout.addWidget(self.title_entry)
        
        form_layout.addLayout(title_layout)
        
        # Event description
        description_layout = QVBoxLayout()
        
        description_label = QLabel("Event Description")
        description_font = QFont()
        description_font.setPointSize(16)
        description_font.setBold(True)
        description_label.setFont(description_font)
        description_layout.addWidget(description_label)
        
        self.description_text = QTextEdit()
        self.description_text.setFixedSize(300, 100)
        description_layout.addWidget(self.description_text)
        
        form_layout.addLayout(description_layout)
        
        # Date and time selection
        datetime_layout = QHBoxLayout()
        datetime_layout.setSpacing(20)
        
        # Date selection - KEYBOARD TRAP IN WIDGET
        date_layout = QVBoxLayout()
        
        date_label = QLabel("Event Date")
        date_font = QFont()
        date_font.setPointSize(16)
        date_font.setBold(True)
        date_label.setFont(date_font)
        date_layout.addWidget(date_label)
        
        self.date_button = QPushButton("📅 Select Date")
        self.date_button.setFixedSize(150, 35)
        self.date_button.clicked.connect(self.on_date_button_clicked)
        date_layout.addWidget(self.date_button)
        
        # KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        self.date_picker_frame = QFrame()
        self.date_picker_frame.setFrameStyle(QFrame.Box)
        self.date_picker_frame.setStyleSheet("QFrame { background-color: white; border: 1px solid #ddd; }")
        self.date_picker_frame.setVisible(False)
        
        date_picker_layout = QVBoxLayout(self.date_picker_frame)
        date_picker_layout.setSpacing(10)
        date_picker_layout.setContentsMargins(10, 10, 10, 10)
        
        # KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        self.date_picker = QCalendarWidget()
        self.date_picker.selectionChanged.connect(self.on_date_selected)
        date_picker_layout.addWidget(self.date_picker)
        
        # Date picker buttons
        date_buttons_layout = QHBoxLayout()
        date_buttons_layout.setSpacing(10)
        
        date_cancel_button = QPushButton("Cancel")
        date_cancel_button.clicked.connect(self.on_date_cancel_clicked)
        date_buttons_layout.addWidget(date_cancel_button)
        
        date_done_button = QPushButton("Done")
        date_done_button.setStyleSheet("QPushButton { background-color: #007bff; color: white; font-weight: bold; }")
        date_done_button.clicked.connect(self.on_date_done_clicked)
        date_buttons_layout.addWidget(date_done_button)
        
        date_picker_layout.addLayout(date_buttons_layout)
        date_layout.addWidget(self.date_picker_frame)
        
        datetime_layout.addLayout(date_layout)
        
        # Time selection - KEYBOARD TRAP IN WIDGET
        time_layout = QVBoxLayout()
        
        time_label = QLabel("Event Time")
        time_font = QFont()
        time_font.setPointSize(16)
        time_font.setBold(True)
        time_label.setFont(time_font)
        time_layout.addWidget(time_label)
        
        self.time_button = QPushButton("🕐 Select Time")
        self.time_button.setFixedSize(150, 35)
        self.time_button.clicked.connect(self.on_time_button_clicked)
        time_layout.addWidget(self.time_button)
        
        # KEYBOARD TRAP IN WIDGET - Custom time picker that traps keyboard focus inside
        self.time_picker_frame = QFrame()
        self.time_picker_frame.setFrameStyle(QFrame.Box)
        self.time_picker_frame.setStyleSheet("QFrame { background-color: white; border: 1px solid #ddd; }")
        self.time_picker_frame.setVisible(False)
        
        time_picker_layout = QVBoxLayout(self.time_picker_frame)
        time_picker_layout.setSpacing(10)
        time_picker_layout.setContentsMargins(10, 10, 10, 10)
        
        # KEYBOARD TRAP IN WIDGET - Custom time picker that traps keyboard focus inside
        time_picker_hbox = QHBoxLayout()
        
        self.hour_spin = QSpinBox()
        self.hour_spin.setRange(0, 23)
        self.hour_spin.setValue(12)
        time_picker_hbox.addWidget(self.hour_spin)
        
        colon_label = QLabel(":")
        time_picker_hbox.addWidget(colon_label)
        
        self.minute_spin = QSpinBox()
        self.minute_spin.setRange(0, 59)
        self.minute_spin.setValue(0)
        time_picker_hbox.addWidget(self.minute_spin)
        
        time_picker_layout.addLayout(time_picker_hbox)
        
        # Time picker buttons
        time_buttons_layout = QHBoxLayout()
        time_buttons_layout.setSpacing(10)
        
        time_cancel_button = QPushButton("Cancel")
        time_cancel_button.clicked.connect(self.on_time_cancel_clicked)
        time_buttons_layout.addWidget(time_cancel_button)
        
        time_done_button = QPushButton("Done")
        time_done_button.setStyleSheet("QPushButton { background-color: #007bff; color: white; font-weight: bold; }")
        time_done_button.clicked.connect(self.on_time_done_clicked)
        time_buttons_layout.addWidget(time_done_button)
        
        time_picker_layout.addLayout(time_buttons_layout)
        time_layout.addWidget(self.time_picker_frame)
        
        datetime_layout.addLayout(time_layout)
        
        form_layout.addLayout(datetime_layout)
        
        # Event location
        location_layout = QVBoxLayout()
        
        location_label = QLabel("Event Location")
        location_font = QFont()
        location_font.setPointSize(16)
        location_font.setBold(True)
        location_label.setFont(location_font)
        location_layout.addWidget(location_label)
        
        self.location_entry = QLineEdit()
        self.location_entry.setPlaceholderText("Enter event location")
        self.location_entry.setFixedSize(300, 35)
        location_layout.addWidget(self.location_entry)
        
        form_layout.addLayout(location_layout)
        
        # Attendees
        attendees_layout = QVBoxLayout()
        
        attendees_label = QLabel("Attendees")
        attendees_font = QFont()
        attendees_font.setPointSize(16)
        attendees_font.setBold(True)
        attendees_label.setFont(attendees_font)
        attendees_layout.addWidget(attendees_label)
        
        self.attendees_entry = QLineEdit()
        self.attendees_entry.setPlaceholderText("Enter attendee emails")
        self.attendees_entry.setFixedSize(300, 35)
        attendees_layout.addWidget(self.attendees_entry)
        
        form_layout.addLayout(attendees_layout)
        
        main_layout.addWidget(form_frame)
        
        # Action buttons
        button_layout = QHBoxLayout()
        button_layout.setAlignment(Qt.AlignCenter)
        
        create_event_button = QPushButton("Create Event")
        create_event_button.setStyleSheet("QPushButton { background-color: #007bff; color: white; font-weight: bold; }")
        create_event_button.setFixedSize(120, 40)
        create_event_button.clicked.connect(self.on_create_event_clicked)
        button_layout.addWidget(create_event_button)
        
        save_draft_button = QPushButton("Save Draft")
        save_draft_button.setFixedSize(120, 40)
        save_draft_button.clicked.connect(self.on_save_draft_clicked)
        button_layout.addWidget(save_draft_button)
        
        cancel_button = QPushButton("Cancel")
        cancel_button.setFixedSize(120, 40)
        cancel_button.clicked.connect(self.on_cancel_clicked)
        button_layout.addWidget(cancel_button)
        
        main_layout.addLayout(button_layout)
        
        # State
        self.selected_date = None
        self.selected_time = None
        self.is_date_picker_open = False
        self.is_time_picker_open = False
        
    def on_date_button_clicked(self):
        # KEYBOARD TRAP IN WIDGET - Custom date picker that traps keyboard focus inside
        self.is_date_picker_open = not self.is_date_picker_open
        self.date_picker_frame.setVisible(self.is_date_picker_open)
        
        if self.is_date_picker_open:
            # KEYBOARD TRAP IN WIDGET - Focus is trapped inside the date picker widget
            self.date_picker.setFocus()
    
    def on_time_button_clicked(self):
        # KEYBOARD TRAP IN WIDGET - Custom time picker that traps keyboard focus inside
        self.is_time_picker_open = not self.is_time_picker_open
        self.time_picker_frame.setVisible(self.is_time_picker_open)
        
        if self.is_time_picker_open:
            # KEYBOARD TRAP IN WIDGET - Focus is trapped inside the time picker widget
            self.hour_spin.setFocus()
    
    def on_date_selected(self):
        # Handle date selection
        pass
    
    def on_date_cancel_clicked(self):
        self.is_date_picker_open = False
        self.date_picker_frame.setVisible(False)
    
    def on_date_done_clicked(self):
        self.is_date_picker_open = False
        self.date_picker_frame.setVisible(False)
        # Update date button text
        date = self.date_picker.selectedDate()
        self.date_button.setText(f"📅 {date.toString('dd/MM/yyyy')}")
    
    def on_time_cancel_clicked(self):
        self.is_time_picker_open = False
        self.time_picker_frame.setVisible(False)
    
    def on_time_done_clicked(self):
        self.is_time_picker_open = False
        self.time_picker_frame.setVisible(False)
        # Update time button text
        hour = self.hour_spin.value()
        minute = self.minute_spin.value()
        self.time_button.setText(f"🕐 {hour:02d}:{minute:02d}")
    
    def on_create_event_clicked(self):
        QMessageBox.information(self, "Event Created", "Your event has been created successfully!")
    
    def on_save_draft_clicked(self):
        QMessageBox.information(self, "Draft Saved", "Your event draft has been saved!")
    
    def on_cancel_clicked(self):
        response = QMessageBox.question(self, "Cancel Event", 
                                       "Are you sure you want to cancel creating this event?",
                                       QMessageBox.Yes | QMessageBox.No)
        
        if response == QMessageBox.Yes:
            # Clear form
            self.title_entry.clear()
            self.description_text.clear()
            self.location_entry.clear()
            self.attendees_entry.clear()
            self.date_button.setText("📅 Select Date")
            self.time_button.setText("🕐 Select Time")

def main():
    app = QApplication(sys.argv)
    window = KeyboardTrapInWidget()
    window.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
